src/java/org/apache/cassandra/index/sai/StorageAttachedIndex.java

```diff
@@ -709,15 +524,164 @@ public IndexBuildingSupport getBuildTaskSupport()
         return INDEX_BUILDER_SUPPORT;
     }
 
-    public IndexContext getIndexContext()
+    /**
+     * Splits SSTables into groups of similar overall size.
+     *
+     * @param toRebuild a list of SSTables to split (Note that this list will be sorted in place!)
+     * @param parallelism an upper bound on the number of groups
+     *
+     * @return a {@link List} of SSTable groups, each represented as a {@link List} of {@link SSTableReader}
+     */
+    @VisibleForTesting
+    public static List<List<SSTableReader>> groupBySize(List<SSTableReader> toRebuild, int parallelism)
     {
-        return indexContext;
+        List<List<SSTableReader>> groups = new ArrayList<>();
+
+        toRebuild.sort(Comparator.comparingLong(SSTableReader::onDiskLength).reversed());
+        Iterator<SSTableReader> sortedSSTables = toRebuild.iterator();
+        double dataPerCompactor = toRebuild.stream().mapToLong(SSTableReader::onDiskLength).sum() * 1.0 / parallelism;
+
+        while (sortedSSTables.hasNext())
+        {
+            long sum = 0;
+            List<SSTableReader> current = new ArrayList<>();
+
+            while (sortedSSTables.hasNext() && sum < dataPerCompactor)
+            {
+                SSTableReader sstable = sortedSSTables.next();
+                sum += sstable.onDiskLength();
+                current.add(sstable);
+            }
+
+            assert !current.isEmpty();
+            groups.add(current);
+        }
+
+        return groups;
     }
 
-    @Override
-    public String toString()
+    /**
+     * @return A set of SSTables which have attached to them invalid index components.
+     */
+    public Collection<SSTableContext> onSSTableChanged(Collection<SSTableReader> oldSSTables, Collection<SSTableContext> newSSTables, IndexValidation validation)
+    {
+        return viewManager.update(oldSSTables, newSSTables, validation);
+    }
+
+    public void drop(Collection<SSTableReader> sstablesToRebuild)
+    {
+        viewManager.drop(sstablesToRebuild);
+    }
+
+    public MemtableIndexManager memtableIndexManager()
+    {
+        return memtableIndexManager;
+    }
+
+    public View view()
+    {
+        return viewManager.view();
+    }
+
+    public IndexTermType indexTermType()
+    {
+        return this.indexTermType;
+    }
+
+    public IndexIdentifier indexIdentifier()
```

Comment:
```suggestion
    public IndexIdentifier identifier()
```
nit: This might look nicer when we have usages like `index.identifier()` rather than `index.indexIdentifier()`...